src/node_api.cc

```diff
@@ -3514,3 +3514,195 @@ napi_status napi_run_script(napi_env env,
   *result = v8impl::JsValueFromV8LocalValue(script_result.ToLocalChecked());
   return GET_RETURN_STATUS(env);
 }
+
+struct napi_threadsafe_function__ {
+  uv_async_t async;
+  napi_ref ref;
+  napi_env env;
+  size_t argc;
+  void* data;
+  void* context;
+  napi_threadsafe_function_marshal marshal_cb;
+  napi_threadsafe_function_process_result process_result_cb;
+};
+
+static napi_value napi_threadsafe_function_error(napi_env env,
+                                                 const char* message) {
+  napi_value result, js_message;
+  if (napi_create_string_utf8(env, message, NAPI_AUTO_LENGTH, &js_message)
+      == napi_ok) {
+    if (napi_create_error(env, nullptr, js_message, &result) == napi_ok) {
+      return result;
+    }
+  }
+
+  napi_fatal_error("N-API async function", NAPI_AUTO_LENGTH,
+      (std::string("Failed to create JS error: ") +
+      std::string(message)).c_str(), NAPI_AUTO_LENGTH);
+  return nullptr;
+}
+
+static void napi_threadsafe_function_cb(uv_async_t* uv_async) {
+  napi_threadsafe_function async =
+      reinterpret_cast<napi_threadsafe_function>(uv_async);
+  napi_handle_scope handle_scope;
+  napi_status status = napi_open_handle_scope(async->env, &handle_scope);
+  if (status == napi_ok) {
+    napi_value js_cb, recv, argv[async->argc], js_result = nullptr,
```

Comment:
VLAs are not 100% portable and prone to blow up the stack when `argc` is large.  You should probably use a std::vector here.

Style: each definition should go on its own line (and separated with `;`, not `,`.)